### PR TITLE
PWN-3247 - Fill history screen

### PR DESCRIPTION
**Pull request by davran1001**
## Jira Ticket (Mandatory)

https://p2pvalidator.atlassian.net/browse/3247

## Description of Work (Mandatory)

Fill history screen with transactions and fix bug with bottom navigation fragments recreation

## Screenshots (Optional)

Screenshot #1 | Screenshot #2 | Screenshot #3
:------------:|:-------------:|:------------:
[Paste Pic 1] | [Paste Pic 2] | [Paste Pic 3]

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryFragment.kt [hunk 1]
```diff
@@ -1,19 +1,65 @@
...
+    private val adapter: HistoryAdapter by unsafeLazy {
+        HistoryAdapter(
+            onTransactionClicked = {},
+            onRetryClicked = {}
+        )
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        Timber.tag("________").d("onViewCreate")
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryFragment.kt:38]: strange tag
maybe we do not need it ?)

File: app/src/main/res/layout/fragment_history.xml [hunk 2]
```diff
@@ -23,31 +23,57 @@
...
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="4dp"
+            android:gravity="center"
+            android:text="@string/receive_no_transactions_yet_message"
+            android:textAppearance="@style/WalletTheme.TextAppearance.SemiBold16" />
+
+    </LinearLayout>
 
-    </androidx.swiperefreshlayout.widget.SwipeRefreshLayout>
```
Comment (kabramovp2p) [inline app/src/main/res/layout/fragment_history.xml:45]: maybe you can just disable pullToRefresh ? not just remove it ?

File: app/src/main/java/org/p2p/wallet/history/HistoryModule.kt [hunk 3]
```diff
@@ -91,6 +91,16 @@ object HistoryModule : InjectionModule {
...
+            HistoryPresenter(
+                historyInteractor = get(),
+                tokenKeyProvider = get(),
+                renBtcInteractor = get(),
+                receiveAnalytics = get(),
+                swapAnalytics = get(),
+                analyticsInteractor = get(),
+                sendAnalytics = get()
+            )
+        } bind HistoryContract.Presenter::class
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/HistoryModule.kt:104]:
```suggestion
        factory<HistoryContract.Presenter> {
            HistoryPresenter(
                historyInteractor = get(),
                tokenKeyProvider = get(),
                renBtcInteractor = get(),
                receiveAnalytics = get(),
                swapAnalytics = get(),
                analyticsInteractor = get(),
                sendAnalytics = get()
            )
        }
```

does it work?

File: app/src/main/java/org/p2p/wallet/history/interactor/HistoryInteractor.kt [hunk 4]
```diff
@@ -74,6 +74,40 @@ class HistoryInteractor(
...
+    ): List<HistoryTransaction> {
+        if (forceNetwork) {
+            transactionsLocalRepository.deleteAll()
+        }
+
+        val signatures = rpcSignatureRepository.getConfirmedSignaturesForAddress(
+            userAccountAddress = tokenPublicKey.toPublicKey(),
+            before = lastSignature,
+            limit = limit
+        ).map(SignatureInformationResponse::signature)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/interactor/HistoryInteractor.kt:91]:
```suggestion
        )
        .map(SignatureInformationResponse::signature)
```

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryFragment.kt [hunk 5]
```diff
@@ -1,19 +1,78 @@
...
+    override fun showPagingState(state: PagingState) {
+        adapter.setPagingState(state)
+        binding.shimmerView.isVisible = state == PagingState.InitialLoading
+    }
+
+    override fun showHistory(items: List<HistoryTransaction>) {
+        Timber.tag("_____").d("Loaded items = ${items.size}")
+        adapter.setTransactions(items)
+
+        val isEmpty = items.isEmpty()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryFragment.kt:62]: what specificly is empty? 
how bout `isEmptyState` `noItemsForList`?

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt [hunk 6]
```diff
@@ -1,5 +1,130 @@
...
+    }
+
+    override fun loadHistory(isRefresh: Boolean) {
+        launch {
+            if (transactions.isEmpty()) {
+                view?.showPagingState(PagingState.InitialLoading)
+            } else {
+                view?.showPagingState(PagingState.Loading)
+            }
+            val lastSignature = transactions.lastOrNull()?.signature
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:45]:
```suggestion
            val lastLoadedSignature = transactions.lastOrNull()?.signature
```

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt [hunk 7]
```diff
@@ -1,5 +1,130 @@
...
+            runCatching {
+                historyInteractor.getTransactionHistory2(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:47]: don't forget to rename

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt [hunk 8]
```diff
@@ -1,5 +1,130 @@
...
+    }
+
+    private fun onSuccess(items: List<HistoryTransaction>) {
+        transactions = transactions + items
+    }
+
+    private fun onFailure(e: Throwable) {
+        Timber.e("Error getting transaction history")
+        if (e is EmptyDataException) {
+            transactions = transactions.toMutableList()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:125]: wait, is this even valid? transactions = emptyList()?

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt [hunk 9]
```diff
@@ -1,5 +1,130 @@
...
+                    // do nothing yet
+                    return@launch
+                }
+            }
+
+            view?.openTransactionDetailsScreen(transaction)
+        }
+    }
+
+    private fun onSuccess(items: List<HistoryTransaction>) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:118]: onTransactionsLoadedSucess

we can onSuccess for different actions, so it's better to be explicit about it

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt [hunk 10]
```diff
@@ -1,5 +1,130 @@
...
+        view?.showHistory(items = newValue)
+        view?.showPagingState(PagingState.Idle)
+    }
+
+    override fun loadHistory(isRefresh: Boolean) {
+        launch {
+            if (transactions.isEmpty()) {
+                view?.showPagingState(PagingState.InitialLoading)
+            } else {
+                view?.showPagingState(PagingState.Loading)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:43]: this PagingState then can be renamed to PagingState.LoadingMore to make that we load more then just loading from the scratch

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt [hunk 11]
```diff
@@ -1,5 +1,130 @@
...
+    private val historyInteractor: HistoryInteractor,
+    private val tokenKeyProvider: TokenKeyProvider,
+    private val renBtcInteractor: RenBtcInteractor,
+    private val receiveAnalytics: ReceiveAnalytics,
+    private val swapAnalytics: SwapAnalytics,
+    private val analyticsInteractor: ScreensAnalyticsInteractor,
+    private val sendAnalytics: SendAnalytics
+) : BasePresenter<HistoryContract.View>(), HistoryContract.Presenter {
+
+    private val publicKey = tokenKeyProvider.publicKey
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:31]: who's public key is that? User's or ...?

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt [hunk 12]
```diff
@@ -1,5 +1,130 @@
...
+        launch {
+            if (transactions.isEmpty()) {
+                view?.showPagingState(PagingState.InitialLoading)
+            } else {
+                view?.showPagingState(PagingState.Loading)
+            }
+            val lastSignature = transactions.lastOrNull()?.signature
+            runCatching {
+                historyInteractor.getTransactionHistory2(
+                    tokenPublicKey = publicKey,
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:48]: we can get tokenPublicKey inside interactor and remove `tokenKeyProvider` from presenter

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt (hunk 6 above)
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:45]: looks like business logic to me, do you mind moving it inside interactor?

File: app/src/main/java/org/p2p/wallet/history/HistoryModule.kt (hunk 3 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/history/HistoryModule.kt:104]: why not ?

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryFragment.kt (hunk 1 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/history/ui/history/HistoryFragment.kt:39]: yeah i just test fragment creation while i switch tabs

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt (hunk 11 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:31]: When you pass registration you create SOL Account so public key it's SOL public key

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt (hunk 6 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:45]: yeah it's business logic, that's why i wrote it in presenter x D

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt (hunk 8 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:125]: No we cannot do so, cause we can reach this exception after fetching first page as examples, i think here will be some boolean flag

File: app/src/main/java/org/p2p/wallet/history/HistoryModule.kt (hunk 3 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/history/HistoryModule.kt:104]: idk, im just suggesting

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt (hunk 6 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt:45]: but business logic should be only in interactor
presenter is responsible only for presentation logic

File: app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt (hunk 10 above)
Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/history/ui/history/HistoryPresenter.kt]: I resolve it with adding param for state

**Review by kabramovp2p — APPROVED** [review]

**Pull request merged**